File: android/src/main/kotlin/com/ax9labs/mapbox_navigation_flutter_v3/mapbox_navigation_flutter_v3/TurnByTurnTheme.kt
package com.ax9labs.mapbox_navigation_flutter_v3.mapbox_navigation_flutter_v3

import org.json.JSONObject

/**
 * Color/shape customization for the turn-by-turn UI, decoded from Dart's
 * `TurnByTurnTheme` (see `lib/src/models.dart`). Every field is nullable -
 * `null` means "use [NavigationTheme.resolveDefaults]'s built-in default"
 * rather than every consumer having to specify a complete palette.
 *
 * Colors are packed ARGB ints - the same representation as Flutter's
 * `Color.value`, so no conversion is needed on either side of the channel.
 */
internal data class TurnByTurnTheme(
    val primaryInstructionColor: Int?,
    val primaryInstructionForegroundColor: Int?,
    val secondaryInstructionColor: Int?,
    val secondaryInstructionForegroundColor: Int?,
    val surfaceColor: Int?,
    val surfaceForegroundColor: Int?,
    val routeColor: Int?,
    val routeCasingColor: Int?,
    val trafficLowColor: Int?,
    val trafficModerateColor: Int?,
    val trafficHeavyColor: Int?,
    val trafficSevereColor: Int?,
    val dividerColor: Int?,
    val secondaryTextColor: Int?,
    val maneuverIconColor: Int?,
    val sheetHandleColor: Int?,
    val cornerRadius: Double?,
    val elevation: Double?
) {
    companion object {
        val DEFAULT =
            TurnByTurnTheme(
                primaryInstructionColor = null,
                primaryInstructionForegroundColor = null,
                secondaryInstructionColor = null,
                secondaryInstructionForegroundColor = null,
                surfaceColor = null,
                surfaceForegroundColor = null,
                routeColor = null,
                routeCasingColor = null,
                trafficLowColor = null,
                trafficModerateColor = null,
                trafficHeavyColor = null,
                trafficSevereColor = null,
                dividerColor = null,
                secondaryTextColor = null,
                maneuverIconColor = null,
                sheetHandleColor = null,
                cornerRadius = null,
                elevation = null
            )

        fun decode(json: JSONObject?): TurnByTurnTheme {
            if (json == null) return DEFAULT
            fun color(key: String): Int? = if (json.has(key) && !json.isNull(key)) json.optInt(key) else null
            fun dimen(key: String): Double? = if (json.has(key) && !json.isNull(key)) json.optDouble(key) else null
            return TurnByTurnTheme(
                primaryInstructionColor = color("primaryInstructionColor"),
                primaryInstructionForegroundColor = color("primaryInstructionForegroundColor"),
                secondaryInstructionColor = color("secondaryInstructionColor"),
                secondaryInstructionForegroundColor = color("secondaryInstructionForegroundColor"),
                surfaceColor = color("surfaceColor"),
                surfaceForegroundColor = color("surfaceForegroundColor"),
                routeColor = color("routeColor"),
                routeCasingColor = color("routeCasingColor"),
                trafficLowColor = color("trafficLowColor"),
                trafficModerateColor = color("trafficModerateColor"),
                trafficHeavyColor = color("trafficHeavyColor"),
                trafficSevereColor = color("trafficSevereColor"),
                dividerColor = color("dividerColor"),
                secondaryTextColor = color("secondaryTextColor"),
                maneuverIconColor = color("maneuverIconColor"),
                sheetHandleColor = color("sheetHandleColor"),
                cornerRadius = dimen("cornerRadius"),
                elevation = dimen("elevation")
            )
        }
    }
}

/**
 * Feature toggles for the turn-by-turn UI, decoded from Dart's
 * `TurnByTurnUiOptions`. See that class's doc comments for what each flag
 * means - kept in sync manually since these mirror each other across the
 * platform channel rather than being generated.
 */
internal data class TurnByTurnUiOptions(
    val showTopBanner: Boolean,
    val showNextManeuver: Boolean,
    val showBottomSheet: Boolean,
    val showTraffic: Boolean,
    val showTrafficSignals: Boolean,
    val showRoadLabels: Boolean,
    val showRecenterButton: Boolean,
    val showRouteOverviewButton: Boolean,
    val enableExpandableBottomSheet: Boolean,
    val confirmBeforeExitNavigation: Boolean,
    val showArrivalTime: Boolean,
    val showRemainingDistance: Boolean,
    val showRemainingDuration: Boolean,
    val enableNavigationCamera: Boolean,
    val enableManeuverAnimations: Boolean
) {
    companion object {
        val DEFAULT =
            TurnByTurnUiOptions(
                showTopBanner = true,
                showNextManeuver = true,
                showBottomSheet = true,
                showTraffic = true,
                showTrafficSignals = false,
                showRoadLabels = true,
                showRecenterButton = true,
                showRouteOverviewButton = true,
                enableExpandableBottomSheet = true,
                confirmBeforeExitNavigation = false,
                showArrivalTime = true,
                showRemainingDistance = true,
                showRemainingDuration = true,
                enableNavigationCamera = true,
                enableManeuverAnimations = true
            )

        fun decode(json: JSONObject?): TurnByTurnUiOptions {
            if (json == null) return DEFAULT
            val d = DEFAULT
            return TurnByTurnUiOptions(
                showTopBanner = json.optBoolean("showTopBanner", d.showTopBanner),
                showNextManeuver = json.optBoolean("showNextManeuver", d.showNextManeuver),
                showBottomSheet = json.optBoolean("showBottomSheet", d.showBottomSheet),
                showTraffic = json.optBoolean("showTraffic", d.showTraffic),
                showTrafficSignals = json.optBoolean("showTrafficSignals", d.showTrafficSignals),
                showRoadLabels = json.optBoolean("showRoadLabels", d.showRoadLabels),
                showRecenterButton = json.optBoolean("showRecenterButton", d.showRecenterButton),
                showRouteOverviewButton = json.optBoolean("showRouteOverviewButton", d.showRouteOverviewButton),
                enableExpandableBottomSheet = json.optBoolean("enableExpandableBottomSheet", d.enableExpandableBottomSheet),
                confirmBeforeExitNavigation = json.optBoolean("confirmBeforeExitNavigation", d.confirmBeforeExitNavigation),
                showArrivalTime = json.optBoolean("showArrivalTime", d.showArrivalTime),
                showRemainingDistance = json.optBoolean("showRemainingDistance", d.showRemainingDistance),
                showRemainingDuration = json.optBoolean("showRemainingDuration", d.showRemainingDuration),
                enableNavigationCamera = json.optBoolean("enableNavigationCamera", d.enableNavigationCamera),
                enableManeuverAnimations = json.optBoolean("enableManeuverAnimations", d.enableManeuverAnimations)
            )
        }
    }
}

File: ios/mapbox_navigation_flutter_v3/Sources/mapbox_navigation_flutter_v3/TurnByTurnTheme.swift
import UIKit

/// Color/shape customization for the turn-by-turn UI, decoded from Dart's
/// `TurnByTurnTheme` (see `lib/src/models.dart`). Every field is optional -
/// `nil` means "use Mapbox's/this plugin's built-in default" rather than
/// every consumer having to specify a complete palette.
///
/// Colors arrive as packed ARGB ints (same representation as Flutter's
/// `Color.value`) - `uiColor` converts one to a `UIColor` when present.
struct TurnByTurnTheme {
    var primaryInstructionColor: Int?
    var primaryInstructionForegroundColor: Int?
    var secondaryInstructionColor: Int?
    var secondaryInstructionForegroundColor: Int?
    var surfaceColor: Int?
    var surfaceForegroundColor: Int?
    var routeColor: Int?
    var routeCasingColor: Int?
    var trafficLowColor: Int?
    var trafficModerateColor: Int?
    var trafficHeavyColor: Int?
    var trafficSevereColor: Int?
    var dividerColor: Int?
    var secondaryTextColor: Int?
    var maneuverIconColor: Int?
    var sheetHandleColor: Int?
    var cornerRadius: Double?
    var elevation: Double?

    static let `default` = TurnByTurnTheme()

    static func decode(_ json: [String: Any]?) -> TurnByTurnTheme {
        guard let json else { return .default }
        var theme = TurnByTurnTheme()
        theme.primaryInstructionColor = json["primaryInstructionColor"] as? Int
        theme.primaryInstructionForegroundColor = json["primaryInstructionForegroundColor"] as? Int
        theme.secondaryInstructionColor = json["secondaryInstructionColor"] as? Int
        theme.secondaryInstructionForegroundColor = json["secondaryInstructionForegroundColor"] as? Int
        theme.surfaceColor = json["surfaceColor"] as? Int
        theme.surfaceForegroundColor = json["surfaceForegroundColor"] as? Int
        theme.routeColor = json["routeColor"] as? Int
        theme.routeCasingColor = json["routeCasingColor"] as? Int
        theme.trafficLowColor = json["trafficLowColor"] as? Int
        theme.trafficModerateColor = json["trafficModerateColor"] as? Int
        theme.trafficHeavyColor = json["trafficHeavyColor"] as? Int
        theme.trafficSevereColor = json["trafficSevereColor"] as? Int
        theme.dividerColor = json["dividerColor"] as? Int
        theme.secondaryTextColor = json["secondaryTextColor"] as? Int
        theme.maneuverIconColor = json["maneuverIconColor"] as? Int
        theme.sheetHandleColor = json["sheetHandleColor"] as? Int
        theme.cornerRadius = json["cornerRadius"] as? Double
        theme.elevation = json["elevation"] as? Double
        return theme
    }
}

extension Int {
    /// Interprets this `Int` as a packed ARGB color (Flutter `Color.value`
    /// representation) and converts it to a `UIColor`.
    var asARGBColor: UIColor {
        let a = CGFloat((self >> 24) & 0xFF) / 255.0
        let r = CGFloat((self >> 16) & 0xFF) / 255.0
        let g = CGFloat((self >> 8) & 0xFF) / 255.0
        let b = CGFloat(self & 0xFF) / 255.0
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}

/// Feature toggles for the turn-by-turn UI, decoded from Dart's
/// `TurnByTurnUiOptions`. Mirrors the Android `TurnByTurnUiOptions.kt`
/// decoder - see that file's doc comments for what each flag means.
struct TurnByTurnUiOptions {
    var showTopBanner = true
    var showNextManeuver = true
    var showBottomSheet = true
    var showTraffic = true
    var showTrafficSignals = false
    var showRoadLabels = true
    var showRecenterButton = true
    var showRouteOverviewButton = true
    var enableExpandableBottomSheet = true
    var confirmBeforeExitNavigation = false
    var showArrivalTime = true
    var showRemainingDistance = true
    var showRemainingDuration = true
    var enableNavigationCamera = true
    var enableManeuverAnimations = true

    static let `default` = TurnByTurnUiOptions()

    static func decode(_ json: [String: Any]?) -> TurnByTurnUiOptions {
        guard let json else { return .default }
        var options = TurnByTurnUiOptions()
        options.showTopBanner = json["showTopBanner"] as? Bool ?? options.showTopBanner
        options.showNextManeuver = json["showNextManeuver"] as? Bool ?? options.showNextManeuver
        options.showBottomSheet = json["showBottomSheet"] as? Bool ?? options.showBottomSheet
        options.showTraffic = json["showTraffic"] as? Bool ?? options.showTraffic
        options.showTrafficSignals = json["showTrafficSignals"] as? Bool ?? options.showTrafficSignals
        options.showRoadLabels = json["showRoadLabels"] as? Bool ?? options.showRoadLabels
        options.showRecenterButton = json["showRecenterButton"] as? Bool ?? options.showRecenterButton
        options.showRouteOverviewButton = json["showRouteOverviewButton"] as? Bool ?? options.showRouteOverviewButton
        options.enableExpandableBottomSheet = json["enableExpandableBottomSheet"] as? Bool ?? options.enableExpandableBottomSheet
        options.confirmBeforeExitNavigation = json["confirmBeforeExitNavigation"] as? Bool ?? options.confirmBeforeExitNavigation
        options.showArrivalTime = json["showArrivalTime"] as? Bool ?? options.showArrivalTime
        options.showRemainingDistance = json["showRemainingDistance"] as? Bool ?? options.showRemainingDistance
        options.showRemainingDuration = json["showRemainingDuration"] as? Bool ?? options.showRemainingDuration
        options.enableNavigationCamera = json["enableNavigationCamera"] as? Bool ?? options.enableNavigationCamera
        options.enableManeuverAnimations = json["enableManeuverAnimations"] as? Bool ?? options.enableManeuverAnimations
        return options
    }
}
